// @bsv/sdk/dist/cjs/src/wallet/substrates/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactNativeWebView = exports.HTTPWalletJSON = exports.HTTPWalletWire = exports.WalletWireProcessor = exports.WalletWireTransceiver = exports.XDM = exports.WindowCWISubstrate = void 0;
var window_CWI_js_1 = require("./window.CWI.js");
Object.defineProperty(exports, "WindowCWISubstrate", { enumerable: true, get: function () { return __importDefault(window_CWI_js_1).default; } });
var XDM_js_1 = require("./XDM.js");
Object.defineProperty(exports, "XDM", { enumerable: true, get: function () { return __importDefault(XDM_js_1).default; } });
__exportStar(require("./WalletWire.js"), exports);
__exportStar(require("./WalletWireCalls.js"), exports);
var WalletWireTransceiver_js_1 = require("./WalletWireTransceiver.js");
Object.defineProperty(exports, "WalletWireTransceiver", { enumerable: true, get: function () { return __importDefault(WalletWireTransceiver_js_1).default; } });
var WalletWireProcessor_js_1 = require("./WalletWireProcessor.js");
Object.defineProperty(exports, "WalletWireProcessor", { enumerable: true, get: function () { return __importDefault(WalletWireProcessor_js_1).default; } });
var HTTPWalletWire_js_1 = require("./HTTPWalletWire.js");
Object.defineProperty(exports, "HTTPWalletWire", { enumerable: true, get: function () { return __importDefault(HTTPWalletWire_js_1).default; } });
var HTTPWalletJSON_js_1 = require("./HTTPWalletJSON.js");
Object.defineProperty(exports, "HTTPWalletJSON", { enumerable: true, get: function () { return __importDefault(HTTPWalletJSON_js_1).default; } });
var ReactNativeWebView_js_1 = require("./ReactNativeWebView.js");
Object.defineProperty(exports, "ReactNativeWebView", { enumerable: true, get: function () { return __importDefault(ReactNativeWebView_js_1).default; } });
