// @bsv/wallet-toolbox-client/out/src/monitor/tasks/TaskPurge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskPurge = void 0;
const WalletMonitorTask_1 = require("./WalletMonitorTask");
class TaskPurge extends WalletMonitorTask_1.WalletMonitorTask {
    constructor(monitor, params, triggerMsecs = 0) {
        super(monitor, TaskPurge.taskName);
        this.params = params;
        this.triggerMsecs = triggerMsecs;
    }
    trigger(nowMsecsSinceEpoch) {
        return {
            run: TaskPurge.checkNow ||
                (this.triggerMsecs > 0 && nowMsecsSinceEpoch - this.lastRunMsecsSinceEpoch > this.triggerMsecs)
        };
    }
    async runTask() {
        let log = '';
        TaskPurge.checkNow = false;
        const r = await this.storage.runAsStorageProvider(async (sp) => {
            return await sp.purgeData(this.params);
        });
        if (r.count > 0)
            log = `${r.count} records updated or deleted.\n${r.log}`;
        return log;
    }
}
exports.TaskPurge = TaskPurge;
TaskPurge.taskName = 'Purge';
/**
 * Set to true to trigger running this task
 */
TaskPurge.checkNow = false;
