// @bsv/wallet-toolbox-client/out/src/storage/schema/entities/EntityOutputTagMap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityOutputTagMap = void 0;
const index_client_1 = require("../../../index.client");
const _1 = require(".");
class EntityOutputTagMap extends _1.EntityBase {
    constructor(api) {
        const now = new Date();
        super(api || {
            created_at: now,
            updated_at: now,
            outputId: 0,
            outputTagId: 0,
            isDeleted: false
        });
    }
    updateApi() {
        /* nothing needed yet... */
    }
    get outputTagId() {
        return this.api.outputTagId;
    }
    set outputTagId(v) {
        this.api.outputTagId = v;
    }
    get outputId() {
        return this.api.outputId;
    }
    set outputId(v) {
        this.api.outputId = v;
    }
    get created_at() {
        return this.api.created_at;
    }
    set created_at(v) {
        this.api.created_at = v;
    }
    get updated_at() {
        return this.api.updated_at;
    }
    set updated_at(v) {
        this.api.updated_at = v;
    }
    get isDeleted() {
        return this.api.isDeleted;
    }
    set isDeleted(v) {
        this.api.isDeleted = v;
    }
    get id() {
        throw new index_client_1.sdk.WERR_INVALID_OPERATION('entity has no "id" value');
    }
    get entityName() {
        return 'outputTagMap';
    }
    get entityTable() {
        return 'output_tags_map';
    }
    equals(ei, syncMap) {
        const eo = this.toApi();
        if (eo.outputId !== (syncMap ? syncMap.output.idMap[(0, index_client_1.verifyId)(ei.outputId)] : ei.outputId) ||
            eo.outputTagId !== (syncMap ? syncMap.outputTag.idMap[(0, index_client_1.verifyId)(ei.outputTagId)] : ei.outputTagId) ||
            eo.isDeleted !== ei.isDeleted)
            return false;
        return true;
    }
    static async mergeFind(storage, userId, ei, syncMap, trx) {
        const outputId = syncMap.output.idMap[ei.outputId];
        const outputTagId = syncMap.outputTag.idMap[ei.outputTagId];
        const ef = (0, index_client_1.verifyOneOrNone)(await storage.findOutputTagMaps({
            partial: { outputId, outputTagId },
            trx
        }));
        return {
            found: !!ef,
            eo: new EntityOutputTagMap(ef || { ...ei }),
            eiId: -1
        };
    }
    async mergeNew(storage, userId, syncMap, trx) {
        this.outputId = syncMap.output.idMap[this.outputId];
        this.outputTagId = syncMap.outputTag.idMap[this.outputTagId];
        await storage.insertOutputTagMap(this.toApi(), trx);
    }
    async mergeExisting(storage, since, ei, syncMap, trx) {
        let wasMerged = false;
        if (ei.updated_at > this.updated_at) {
            this.isDeleted = ei.isDeleted;
            this.updated_at = new Date(Math.max(ei.updated_at.getTime(), this.updated_at.getTime()));
            await storage.updateOutputTagMap(this.outputId, this.outputTagId, this.toApi(), trx);
            wasMerged = true;
        }
        return wasMerged;
    }
}
exports.EntityOutputTagMap = EntityOutputTagMap;
