// @bsv/wallet-toolbox-client/out/src/storage/schema/entities/EntityOutputTag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityOutputTag = void 0;
const index_client_1 = require("../../../index.client");
const _1 = require(".");
class EntityOutputTag extends _1.EntityBase {
    constructor(api) {
        const now = new Date();
        super(api || {
            outputTagId: 0,
            created_at: now,
            updated_at: now,
            tag: '',
            userId: 0,
            isDeleted: false
        });
    }
    updateApi() {
        /* nothing needed yet... */
    }
    get outputTagId() {
        return this.api.outputTagId;
    }
    set outputTagId(v) {
        this.api.outputTagId = v;
    }
    get created_at() {
        return this.api.created_at;
    }
    set created_at(v) {
        this.api.created_at = v;
    }
    get updated_at() {
        return this.api.updated_at;
    }
    set updated_at(v) {
        this.api.updated_at = v;
    }
    get tag() {
        return this.api.tag;
    }
    set tag(v) {
        this.api.tag = v;
    }
    get userId() {
        return this.api.userId;
    }
    set userId(v) {
        this.api.userId = v;
    }
    get isDeleted() {
        return this.api.isDeleted;
    }
    set isDeleted(v) {
        this.api.isDeleted = v;
    }
    get id() {
        return this.api.outputTagId;
    }
    set id(v) {
        this.api.outputTagId = v;
    }
    get entityName() {
        return 'outputTag';
    }
    get entityTable() {
        return 'output_tags';
    }
    equals(ei, syncMap) {
        const eo = this.toApi();
        if (eo.tag != ei.tag || eo.isDeleted != ei.isDeleted)
            return false;
        if (!syncMap) {
            if (eo.userId !== ei.userId)
                return false;
        }
        return true;
    }
    static async mergeFind(storage, userId, ei, syncMap, trx) {
        const ef = (0, index_client_1.verifyOneOrNone)(await storage.findOutputTags({ partial: { tag: ei.tag, userId }, trx }));
        return {
            found: !!ef,
            eo: new EntityOutputTag(ef || { ...ei }),
            eiId: (0, index_client_1.verifyId)(ei.outputTagId)
        };
    }
    async mergeNew(storage, userId, syncMap, trx) {
        this.userId = userId;
        this.outputTagId = 0;
        this.outputTagId = await storage.insertOutputTag(this.toApi(), trx);
    }
    async mergeExisting(storage, since, ei, syncMap, trx) {
        let wasMerged = false;
        if (ei.updated_at > this.updated_at) {
            this.isDeleted = ei.isDeleted;
            this.updated_at = new Date(Math.max(ei.updated_at.getTime(), this.updated_at.getTime()));
            await storage.updateOutputTag(this.id, this.toApi(), trx);
            wasMerged = true;
        }
        return wasMerged;
    }
}
exports.EntityOutputTag = EntityOutputTag;
