// @bsv/sdk/dist/cjs/src/wallet/WERR_REVIEW_ACTIONS.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WERR_REVIEW_ACTIONS = void 0;
/**
 * When a `createAction` or `signAction` is completed in undelayed mode (`acceptDelayedBroadcast`: false),
 * any unsucccessful result will return the results by way of this exception to ensure attention is
 * paid to processing errors.
 */
class WERR_REVIEW_ACTIONS extends Error {
    /**
     * All parameters correspond to their comparable `createAction` or `signSction` results
     * with the exception of `reviewActionResults`;
     * which contains more details, particularly for double spend results.
     */
    constructor(reviewActionResults, sendWithResults, txid, tx, noSendChange) {
        super('Undelayed createAction or signAction results require review.');
        this.reviewActionResults = reviewActionResults;
        this.sendWithResults = sendWithResults;
        this.txid = txid;
        this.tx = tx;
        this.noSendChange = noSendChange;
        this.isError = true;
        this.code = 5;
        this.name = this.constructor.name;
    }
}
exports.WERR_REVIEW_ACTIONS = WERR_REVIEW_ACTIONS;
exports.default = WERR_REVIEW_ACTIONS;
