// @bsv/wallet-toolbox-client/out/src/signer/methods/acquireDirectCertificate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.acquireDirectCertificate = acquireDirectCertificate;
async function acquireDirectCertificate(wallet, auth, vargs) {
    var _a;
    const now = new Date();
    const newCert = {
        certificateId: 0, // replaced by storage insert
        created_at: now,
        updated_at: now,
        userId: auth.userId,
        type: vargs.type,
        subject: vargs.subject,
        verifier: vargs.keyringRevealer === 'certifier' ? vargs.certifier : vargs.keyringRevealer,
        serialNumber: vargs.serialNumber,
        certifier: vargs.certifier,
        revocationOutpoint: vargs.revocationOutpoint,
        signature: vargs.signature,
        fields: [],
        isDeleted: false
    };
    for (const [name, value] of Object.entries(vargs.fields)) {
        (_a = newCert.fields) === null || _a === void 0 ? void 0 : _a.push({
            certificateId: 0, // replaced by storage insert
            created_at: now,
            updated_at: now,
            userId: auth.userId,
            fieldName: name,
            fieldValue: value,
            masterKey: vargs.keyringForSubject[name] || ''
        });
    }
    const count = await wallet.storage.insertCertificate(newCert);
    const r = {
        type: vargs.type,
        subject: vargs.subject,
        serialNumber: vargs.serialNumber,
        certifier: vargs.certifier,
        revocationOutpoint: vargs.revocationOutpoint,
        signature: vargs.signature,
        fields: vargs.fields
    };
    return r;
}
